pkg/worker/worker.go

```diff
@@ -0,0 +1,300 @@
+method.
+
+DataStorage interface must include working data synchronization methods: Lock()
+and Unlock().
+*/
+package worker
+
+import (
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/google/uuid"
+)
+
+var (
+	// ErrJobNotFound indicates that the requested job
+	// has not been found in the DataStorage.
+	ErrJobNotFound = errors.New("job has not been found")
+
+	// ErrJobAlreadyStopped indicates that the requested job
+	// has already been stopped.
+	ErrJobAlreadyStopped = errors.New("job has already been stopped")
+
+	// ErrJobAlreadyFinished indicates that the requested job
+	// has already been finished.
+	ErrJobAlreadyFinished = errors.New("job has already been finished")
+)
+
+// DataStorage allows to store and retrieve jobs from any storage that implements the interface.
+// The storage needs to implement sync.Mutex or other form of access synchronization based on
+// Lock() and Unlock() methods.
+type DataStorage interface {
+	Store(job *Job) error
+	Retrieve(uuid uuid.UUID) (*Job, error)
+}
+
+// MemoryStorage is a simple, in-memory implementation of the DataStorage interface.
+// Avoid using in production. Implement DataStorage with database of your choice.
+type MemoryStorage struct {
+	Data  map[uuid.UUID]*Job
+	mutex sync.Mutex
+}
+
+// Store saves the job in memory.
+func (m *MemoryStorage) Store(job *Job) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	logs := job.Logs()
+	job.SetLogs(logs)
+	m.Data[job.UUID] = job
+	return nil
+}
+
+// Retrieve loads the job from the memory storage.
+func (m *MemoryStorage) Retrieve(uuid uuid.UUID) (*Job, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	job, ok := m.Data[uuid]
+	if !ok {
+		return nil, ErrJobNotFound
+	}
+	logs := job.Logs()
+	job.SetLogs(logs)
+	return job, nil
+}
+
+// Worker provides an interface for running, stopping and querying processes.
+type Worker struct {
+	DataStorage DataStorage
+}
+
+// NewWorker constructs a new Worker with a struct that implements DataStorage interface.
+func NewWorker(dataStorage DataStorage) *Worker {
+	return &Worker{DataStorage: dataStorage}
+}
+
+// Start creates a new Linux process with the provided command,
+// instantiates a new Job object and returns its ID.
+// It runs the job in a goroutine that updates its progress directly
+// to the DataStorage.
+// TODO: Add a proper logger (e.g. logrus) interface to the Worker struct to standarize logging.
+// TODO: Report errors and job completion status via channels.
+// TODO: Add a optional timeout for better control over goroutines.
+//
+// Notes: I generally try to avoid using goroutines in packages. It introduces unpredictable
+// side effects and takes control away from the client. I couldn't come up with a better
+// solution to allow this package to control StdIn and StdErr of the process.
+func (worker *Worker) Start(command []string, user string) uuid.UUID {
```

Comment:
what happens when I call `worker.Start(nil, "")` ?

Reply:
Nice catch. I changed the function return values to also contain errors and added a check of the user length.